sdk/client-core/core/src/main/java/io/clientcore/core/http/client/DefaultHttpClient.java

```diff
@@ -471,38 +469,93 @@ private static void buildAndSend(HttpRequest httpRequest, OutputStreamWriter out
+        private static Response<?> buildResponse(HttpRequest httpRequest, BufferedInputStream inputStream)
+            throws IOException {
+            // Parse Http response from socket:
+            // Status Line
+            // Response Headers
+            // Blank Line
+            // Response Body
+
+            int statusCode = readStatusCode(inputStream);
+            HttpHeaders headers = readResponseHeaders(inputStream);
+            // read body if present
+            // TODO: (add chunked encoding support)
+            HttpHeader contentLengthHeader = headers.get(CONTENT_LENGTH);
+            byte[] body = getBody(inputStream, contentLengthHeader);
+            if (body != null) {
+                return new HttpResponse<>(httpRequest, statusCode, headers, BinaryData.fromBytes(body));
+            }
+            return new HttpResponse<>(httpRequest, statusCode, headers, null);
+        }
 
-            HttpHeaders headers = new HttpHeaders();
-            String line;
-            while ((line = reader.readLine()) != null && !line.isEmpty()) {
-                // Headers may have optional leading and trailing whitespace around the header value.
-                // https://tools.ietf.org/html/rfc7230#section-3.2
-                // Process this accordingly.
-                int split = line.indexOf(':'); // Find ':' to split the header name and value.
-                String key = line.substring(0, split); // Get the header name.
-                String value = line.substring(split + 1).trim(); // Get the header value and trim whitespace.
-                headers.add(HttpHeaderName.fromString(key), value);
+        private static byte[] getBody(BufferedInputStream inputStream, HttpHeader contentLengthHeader)
+            throws IOException {
+            int contentLength;
+            if (contentLengthHeader == null || contentLengthHeader.getValue() == null) {
+                contentLength = -1;
+            } else {
+                contentLength = Integer.parseInt(contentLengthHeader.getValue());
             }
+            if (contentLength > 0) {
+                byte[] buffer = new byte[contentLength];
+                int bytesRead;
+                int totalBytesRead = 0;
+
+                while (totalBytesRead < contentLength
+                    && (bytesRead = inputStream.read(buffer, totalBytesRead, contentLength - totalBytesRead)) != -1) {
+                    totalBytesRead += bytesRead;
+                }
+
+                return buffer;
+            }
+            return null;
+        }
 
-            StringBuilder bodyString = new StringBuilder();
-            while ((line = reader.readLine()) != null) {
-                bodyString.append(line);
+        private static int readStatusCode(InputStream inputStream) throws IOException {
+            ByteArrayOutputStream byteOutputStream = new ByteArrayOutputStream();
+            int b;
+            while ((b = inputStream.read()) != -1 && b != '\n') {
```

Comment:
One thing we may need to look at in the future for performance is reading in chunks, reading byte-by-byte off the network will be costly.

Reply:
Since we are dealing with persistent connections and inputstreams, we should be better off reading byte-by-byte rather than buffering readers or readLines I was thinking.
What could be the alternative otherwise?